### ISSUE TITLE
Banning a virtual/abstract method should also ban usages of overridden versions of it

### ISSUE THREAD

**Issue by davkean** [dotnet/roslyn-analyzers#3295]
The API analyzer does fire on overrides of a method that has been banned, which makes the below pattern very hard to ban.

#### Analyzer package

BannedApiAnalyzers

#### Diagnostic ID

RS0030

#### Repro steps

```
M:Microsoft.VisualStudio.ProjectSystem.OnceInitializedOnceDisposed.Initialize();Did you mean OnceInitializedOnceDisposed.EnsureInitialized?
```

``` C#

public class Component : OnceInitializedOnceDisposed
{
    public string Property
    {
        get 
        {
            Initialized();
            return "Hello";
        }
    }

   protected override void Initialize()
   {
   }
}
```

#### Expected behavior
Warns about incorrect usage of Initialize

#### Actual behavior
Does not warn

**Comment by Evangelink** [comment]
@davkean Do you expect an issue to be raised only on the call to `Initialized` in the getter or also on the override itself?

**Comment by Evangelink** [comment]
@mavasani @davkean Do you think that we should raise a diagnostic on the call to base?

```csharp
public class C1
{
	public void Initialize() {}
}

public class C2
{
	public override Initialize()
	{
		base.Initialize(); // what do we want to do here? IMO there shouldn't be a diag
	}

	public void M()
	{
		Initialize(); // error here makes sense
	}
}
```

**Comment by mavasani** [comment]
@Evangelink I think the original issue raised by @davkean is still not fixed. For example, we should fire a diagnostic on the _declaration site_ `public override void Method1() {}` defined in your test here: https://github.com/dotnet/roslyn-analyzers/pull/3404/files#diff-0feca8f7d5dc5a6c19832e70e82d2b2aR786. We only handled the case where the rule now fires on _call site_, which is necessary but not a sufficient fix.

**Comment by davkean** [comment]
The behavior is as expected apart from below, its firing on the following:

```
        protected override void Initialize()
        {
            base.Initialize(); // Fires
        } 
```

This is unlike the obsolete warning which does not fire on methods that them themselves are obsolete. I don't want to fire on overrides as they are "banned" and hence should be allowed to call themselves or base.

**Comment by Evangelink** [comment]
@mavasani I am ok with raising on the _declaration site_ for `public override void Method2() {}` on `C2` because it's `virtual` on `C1` and banned on `C1` but it seems weird to me to raise on `public override void Method1() {}` of `C2` as it is `abstract` on `C1`. If we do want to raise on declaration of `Method1` why don't we simply mark all `C1` as banned because effectively you can no longer inherit from `C1`.

@davkean I am sorry but I am not sure to fully understand what you meant. We are currently raising on the call to `base.Initialize()` and you would like to avoid so, isn't it? Also for the last part of your sentence, do you mean that it's ok to do a direct recursive call (but still not ok to do indirect one).

```csharp
class C
{
	public int M(int i) // This method is banned
	{
		return M(i -1); // This recursive call is allowed
	}
}
```

**Comment by davkean** [comment]
This analyzer is the equivalent of applying [Obsolete] to an API you don't own and hence we should model on the [Obsolete] analysis:

https://sharplab.io/#v2:D4AQTAjAsAULIGYAEBDARgZwC4CcUGMslwkAxAe3NgG9Yl6kBtAeU3IBsBTLTgXToaJUmXASIgALEgCy3ABbkAJhAAUASgDcsAfRZsuPfjAbFkANwCWOLAFcU7YlNlYFisOqTUkAX1i+4MEIkAEIoOEgAXGSUNDqmjkgAggB25C6cOM6u6nG0xiYMLhYYAHRZSu6aBdXVcSZoKBicZfIV6lr59P5xQuRmGTgWipwJ5coeNQD0k0ihycmcikkACgCSSIrknBhIqUQA7mHJSOTHRRi52p0+VyZ6GBzcfD3IfQNDI5IyrW4513k1JANJotFxtKqAyFIaazFDzRYrdaHHDJHanJDnOL+fxAA

There's actually a separate warning for overriding an obsolete API without obsoleting the override.

**Comment by JakeYallop** [comment]
I've just ran into this issue myself. I started to look at fixing it, however, I ran into an issue with `DocumentationCommentId`. I've opened an issue in the roslyn repo: https://github.com/dotnet/roslyn/issues/64915
